### PR TITLE
Update Runner to send step updates to Results

### PR DESCRIPTION
**Pull request by yacaovsnc**
* Refactored `jobServer` so `Results` code are in `ResultsServer` now.
* Updated `ResultsServer` to send steps update to Results
* Results upload is best effort and any exception should skip subsequent uploads
* Add authenticated proxy support
* Buffer request for `WinHttpHandler`

### CODE REVIEW

File: src/Runner.Worker/JobRunner.cs [hunk 1]
```diff
@@ -49,6 +49,8 @@ public async Task<TaskResult> RunAsync(AgentJobRequestMessage message, Cancellat
                 VssCredentials jobServerCredential = VssUtil.GetVssCredential(systemConnection);
                 await runServer.ConnectAsync(systemConnection.Url, jobServerCredential);
                 server = runServer;
+
+                _jobServerQueue.Start(message);
```
Comment (brittanyellich) [inline src/Runner.Worker/JobRunner.cs:53]: Just verifying that this change should be here as it doesn't look related, but maybe it is?

File: src/Sdk/WebApi/WebApi/Contracts.cs [hunk 2]
```diff
@@ -126,6 +127,46 @@ public class CreateMetadataResponse
...
+        public string Name;
+        [DataMember]
+        public Status Status;
+        [DataMember]
+        public string StartedAt;
+        [DataMember]
+        public string CompletedAt;
+    }
+
+    public enum Status
```
Comment (brittanyellich) [inline src/Sdk/WebApi/WebApi/Contracts.cs:162]: It may be helpful to link to somewhere to indicate why were are choosing these particular status numbers? Otherwise the context on the strange ordering of the numbers may get lost. Unless there is already an explanation in results.

**Review by brittanyellich — APPROVED** [review]

File: src/Runner.Worker/JobRunner.cs (hunk 1 above)
Comment (yacaovsnc) [reply src/Runner.Worker/JobRunner.cs]: Ah, good eye!  I forgot to remove it.  This will be in next PR where we may skip uploading with `JobServer` but need to start `ResultsServer`.

File: src/Sdk/WebApi/WebApi/Contracts.cs (hunk 2 above)
Comment (yacaovsnc) [reply src/Sdk/WebApi/WebApi/Contracts.cs:162]: We have a list of `Status` and the numbers are consecutive, but only those are relevant to steps.  🤷🏼

File: src/Sdk/Common/Common/VssHttpMessageHandler.cs
Comment (TingluoHuang) [inline src/Sdk/Common/Common/VssHttpMessageHandler.cs:1]: undo the change if nothing changed. 😄

File: src/Sdk/Common/Common/VssHttpRequestSettings.cs
Comment (TingluoHuang) [inline src/Sdk/Common/Common/VssHttpRequestSettings.cs:1]: same, undo if not needed.

File: src/Sdk/Common/Common/VssHttpMessageHandler.cs
Comment (yacaovsnc) [reply src/Sdk/Common/Common/VssHttpMessageHandler.cs:1]: Looks like a linter change. I will undo.

File: src/Runner.Common/JobServerQueue.cs [hunk 3]
```diff
@@ -111,8 +113,11 @@ public void Start(Pipelines.AgentJobRequestMessage jobRequest)
                 !string.IsNullOrEmpty(resultsReceiverEndpoint))
             {
                 Trace.Info("Initializing results client");
-                _jobServer.InitializeResultsClient(new Uri(resultsReceiverEndpoint), accessToken);
-                _resultsClientInitiated = true;
+                if (_resultsServer != null)
```
Comment (TingluoHuang) [inline src/Runner.Common/JobServerQueue.cs:116]: `_resultsServer ` should never be NULL since it's created by `public override void Initialize(IHostContext hostContext)`

File: src/Runner.Common/JobServerQueue.cs [hunk 4]
```diff
@@ -612,6 +620,21 @@ private async Task ProcessTimelinesUpdateQueueAsync(bool runOnce = false)
...
                             await _jobServer.UpdateTimelineRecordsAsync(_scopeIdentifier, _hubName, _planId, update.TimelineId, update.PendingRecords, default(CancellationToken));
+                            try
+                            {
+                                if (_resultsClientInitiated)
+                                {
+                                    await _resultsServer.UpdateResultsWorkflowStepsAsync(_scopeIdentifier, _hubName, _planId, update.TimelineId, update.PendingRecords, default(CancellationToken));
+                                }
+                            }
+                            catch (Exception e)
+                            {
```
Comment (TingluoHuang) [inline src/Runner.Common/JobServerQueue.cs:631]: do we want to report the error to the actions service as telemetry? like how we did for log upload?

File: src/Runner.Common/JobServerQueue.cs (hunk 3 above)
Comment (yacaovsnc) [reply src/Runner.Common/JobServerQueue.cs]: Removed null guard.

File: src/Runner.Common/ResultsServer.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using GitHub.DistributedTask.WebApi;
+using GitHub.Services.Results.Client;
+
+namespace GitHub.Runner.Common
+{
+    [ServiceLocator(Default = typeof(ResultServer))]
+    public interface IResultsServer : IRunnerService, IAsyncDisposable
```
Comment (TingluoHuang) [inline src/Runner.Common/ResultsServer.cs:12]: do you need `IAsyncDisposable`?

File: src/Sdk/WebApi/WebApi/Contracts.cs [hunk 6]
```diff
@@ -126,6 +127,46 @@ public class CreateMetadataResponse
...
+    {
+        StatusUnknown = 0,
+        StatusInProgress = 3,
+        StatusPending = 5,
+        StatusCompleted = 6
```
Comment (TingluoHuang) [inline src/Sdk/WebApi/WebApi/Contracts.cs:167]: why not 0, 1, 2, 3?

Comment (yacaovsnc) [reply src/Sdk/WebApi/WebApi/Contracts.cs:167]: Those numbers are defined elsewhere.  I removed the ones that are not applicable to runner.

File: src/Runner.Common/ResultsServer.cs (hunk 5 above)
Comment (yacaovsnc) [reply src/Runner.Common/ResultsServer.cs:12]: I was going to move the web socket client to here in the future so I left this.  I will remove this for now since it's not needed.

**Review by TingluoHuang — APPROVED** [review]
